src/aggregate_verify.cc

```diff
@@ -0,0 +1,226 @@
+#include "functions-inl.h"
+
+namespace blst_ts_functions {
+typedef struct {
+    blst_ts::P1AffineGroup pk_point;
+    uint8_t *msg;
+    size_t msg_len;
+} AggregateVerifySet;
+
+/**
+ * @param[out]  {blst_ts::P2AffineGroup} - P2 point for verification
+ * @param[out]  {std::vector<AggregateVerifySet>} - Sets to be added to pairing
+ * @param[in]   {Napi::CallbackInfo} - JS function context
+ */
+blst_ts::BLST_TS_ERROR prepare_aggregate_verify(
+    blst_ts::P2AffineGroup &sig_point,
+    std::vector<AggregateVerifySet> &sets,
+    const Napi::CallbackInfo &info) {
+    Napi::Env env = info.Env();
+    blst_ts::BLST_TS_ERROR err = unwrap_signature(sig_point, env, info[2]);
+    if (err != blst_ts::BLST_TS_ERROR::SUCCESS) {
+        return err;
+    }
+
+    if (!info[0].IsArray()) {
+        Napi::TypeError::New(
+            env, "BLST_ERROR: msgs must be of type BlstBuffer[]")
+            .ThrowAsJavaScriptException();
+        return blst_ts::BLST_TS_ERROR::JS_ERROR_THROWN;
+    }
+    Napi::Array msgs_array = info[0].As<Napi::Array>();
+    uint32_t msgs_array_length = msgs_array.Length();
+    if (msgs_array_length == 0) {
+        Napi::TypeError::New(env, "BLST_ERROR: msgs must have length > 0")
+            .ThrowAsJavaScriptException();
+        return blst_ts::BLST_TS_ERROR::JS_ERROR_THROWN;
+    }
+
+    if (!info[1].IsArray()) {
+        Napi::TypeError::New(env, "publicKeys must be of type PublicKeyArg[]")
+            .ThrowAsJavaScriptException();
+        return blst_ts::BLST_TS_ERROR::JS_ERROR_THROWN;
+    }
+    Napi::Array pk_array = info[1].As<Napi::Array>();
+    uint32_t pk_array_length = pk_array.Length();
+    if (pk_array_length == 0) {
+        if (sig_point.raw_point->is_inf()) {
+            return blst_ts::BLST_TS_ERROR::INVALID;
+        }
+        Napi::TypeError::New(env, "BLST_ERROR: publicKeys must have length > 0")
+            .ThrowAsJavaScriptException();
+        return blst_ts::BLST_TS_ERROR::JS_ERROR_THROWN;
+    }
+
+    if (msgs_array_length != pk_array_length) {
+        Napi::TypeError::New(
+            env, "BLST_ERROR: msgs and publicKeys must be the same length")
+            .ThrowAsJavaScriptException();
+        return blst_ts::BLST_TS_ERROR::JS_ERROR_THROWN;
+    }
+
+    sets.reserve(pk_array_length);
+    for (uint32_t i = 0; i < pk_array_length; i++) {
+        sets.push_back({blst_ts::P1AffineGroup{}, nullptr, 0});
+
+        Napi::Value msg_value = msgs_array[i];
+        BLST_TS_FUNCTION_UNWRAP_UINT_8_ARRAY(msg_value, msg, "msg");
+        sets[i].msg = msg.Data();
+        sets[i].msg_len = msg.ByteLength();
+
+        err = unwrap_public_key(sets[i].pk_point, env, pk_array[i]);
+        if (err != blst_ts::BLST_TS_ERROR::SUCCESS) {
+            return err;
+        }
+    }
+
+    return blst_ts::BLST_TS_ERROR::SUCCESS;
+}
+
+/**
+ * @param[out]  {boolean} - Result of the verification
```

Comment:
Is there a reason you're using `out` parameters and not returning values? Especially for boolean.

Reply:
Yes.  Need to return a function execution result/status enum to determine what type of failure happened.  Some cases a failure should throw an error and in some cases the failure should just cause the outer function to return `false`

Reply:
Would it be more idiomatic to throw an exception instead? That is, if no exception was thrown, it means all is good. If an exception is thrown, you get the error message.

Reply:
Following that, you instead of using enums to denote what kind of error happened, you could have custom exception classes to do that instead. Given that the code is already throwing exceptions, it would seem to me it would work well in your case.

Reply:
Will look at that in the next refactor.  Its my understanding that throwing is slower than returning values because of the stack unwinds.  Not sure but I'm also not super familiar with custom exceptions.  To be continued at a later date though!